### PR TITLE
Properly position absolutes with static insets that are children of floats

### LINKED ISSUE #29826: Layout 2020: Absolutely positioned elements with a static position are not properly positioned in floats
This can be seen by observing the rendering of [absolute_hypothetical_float.html](https://github.com/servo/servo/blob/1fb7d45d5ff15dd1df67e6dfc62db813d2e2e948/tests/wpt/mozilla/tests/css/absolute_hypothetical_float.html).

cc @Loirooriol

### LINKED ISSUE #29886: Layout 2020: Floats don't take into account that parent may get some margin collapse from children
```html
<!DOCTYPE html>
<div style="outline: solid orange">
  <div style="float: left; width: 25px; height: 25px; background: cyan"></div>
  <div style="margin-top: 50px; width: 50px; height: 50px; background: magenta"></div>
</div>
```

| Layout 2013 (good) | Layout 2020 (bad) |
| ------------- | ------------- |
|   |  |

This is violating https://drafts.csswg.org/css2/#float-rules

> 4. A floating box’s [outer top](https://drafts.csswg.org/css2/#outer-edge) may not be higher than the top of its [containing block](https://drafts.csswg.org/css2/#containing-block).

### PR DESCRIPTION
**Pull request by mrobinson**
Previously, final float positions were calculated when their parents
were positioned. This prevented proper positioning of absolute children
of floats with static insets, because they accumulate offsets as they
are hoisted up the tree.

This change moves the final float positioning to
`PlacementState::place_fragment` for the float itself so that it happens
before any insets are updated for hoisted descendants. In addition to
simplifying the code, this makes it a bit more efficient. Finally,
floats are taken into account when updating static insets of hoisted
boxes.

Fixes #29826.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #29826
- [x] There are tests for these changes

### CODE REVIEW

**Comment by mrobinson** [comment]
@Loirooriol This one is ready to review. Thanks!

**Review by Loirooriol — APPROVED** [review]
Seems reasonable, but you will have to rebase on top of #29902 to avoid a merge conflict. And then I would run the bot again to ensure nothing has changed.

**Comment by Loirooriol** [comment]
This seems to fix #29886 too :sparkles:

File: tests/wpt/mozilla/tests/css/absolute_hypothetical_float.html [hunk 1]
```diff
@@ -1,22 +0,0 @@
-<!DOCTYPE html>
```
[the remark this answers is not in the document]
Comment (mrobinson) [reply tests/wpt/mozilla/tests/css/absolute_hypothetical_float.html:1]: Yeah, the test is a port of the old one into the WPT suite.

File: tests/wpt/web-platform-tests/css/CSS2/floats/float-with-absolutely-positioned-child-with-static-inset-ref.html [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    height: 100px;
+}
+
+#abs-grandchild {
+    background: darkgreen;
+    width: 50px;
+    height: 50px;
+}
+</style>
+</head>
```
[the remark this answers is not in the document]
Comment (mrobinson) [reply tests/wpt/web-platform-tests/css/CSS2/floats/float-with-absolutely-positioned-child-with-static-inset-ref.html:24]: Whoops, looks like this is a copy and paste issue.

**Pull request merged**